### ISSUE TITLE
Fix Note9 crash on startup

### ISSUE THREAD

**Issue by uazo** [bromite/bromite#1871]
### Preliminary checklist

- [X] I have read the [README](https://github.com/bromite/bromite/blob/master/README.md)
- [X] I have read the [FAQs](https://github.com/bromite/bromite/blob/master/FAQ.md).
- [X] I have searched [existing issues](https://github.com/bromite/bromite/issues) for my feature request. This is a new issue (NOT a duplicate) and is not related to another issue.

### Is your feature request related to privacy?

No

### Is there a patch available for this feature somewhere?

[yes](https://github.com/uazo/bromite/blob/next/v99/build/patches/Fix-Note9-startup-crash.patch)

### Describe the solution you would like

as described [here](https://github.com/ungoogled-software/ungoogled-chromium-android/issues/113), if you want you can allow the use of bromite even for users with note9 who have a crash on startup

### Describe alternatives you have considered

none

**Comment by kokukyocho** [comment]
Hello @uazo!

I am a bot :robot:

Thanks for submitting this issue!
Sorry, this is not a feature request related to privacy. It might be considered for development but it is invalid because - as explained in the template - only feature requests related to privacy are accepted.


Additionally, this issue seems to be a duplicate of the following ones:
* #1833

**Comment by csagan5** [comment]
I am not sure about this, there are 2 issues:
* we do not have model-specific fixes in the patches so far
* are there security problems introduced when changing the runtime linker that way? Performance issues?

Also: why does it happen in Chromium but not Chrome? Perhaps there are some special fieldtrials/flags that would solve this?

**Comment by uazo** [comment]
> we do not have model-specific fixes in the patches so far

I know. in fact you decide if you want to add it, it is specific for that model.

> are there security problems introduced when changing the runtime linker that way? 

loosing RELRO, just because is not supported from that specific combination of device / android version.

> Performance issues?

no, but in general the use of chrome_public is not suitable for recent android, because the new native linkers seem to manage the memory better when forking process for the RELRO.

> Also: why does it happen in Chromium but not Chrome?

I don't know, based on what is communicated by the chromium team, it seems that google play will send, if necessary, a specific version based on the model, but beyond that we do not know.
in chromium it is known to not work, as written in https://groups.google.com/a/chromium.org/g/chromium-dev/c/iAb7QUiNPLw

> Perhaps there are some special fieldtrials/flags that would solve this?

i checked, but i didn't find anything.

**Comment by csagan5** [comment]
> > Performance issues?
> 
> no, but in general the use of chrome_public is not suitable for recent android, because the new native linkers seem to manage the memory better when forking process for the RELRO.

Is there an alternative for Bromite? As far as I remember the other targets (like Vanadium also does) would require something to be embedded in the ROM.

**Comment by uazo** [comment]
> Is there an alternative for Bromite?

monochrome or trichrome, but it would force us to create more versions by type of device, perhaps without major advantages but I don't know. or try to force the use of the native linker (like that patch but on the contrary) but i read the linker code, for me it is arabic.

**Comment by csagan5** [comment]
As far as I know monochrome or trichrome have 0 advantages if you do not also modify the SystemWebView in the ROM, because they have advantages only when that is loaded into memory.

> create more versions by type of device

Can you make an example?

> or try to force the use of the system linker (like that patch but on the contrary) but i read the linker code, for me it is arabic.

If there is no Bromite SystemWebView, this will also do nothing?

**Comment by uazo** [comment]
> As far as I know monochrome or trichrome have 0 advantages if you do not also modify the SystemWebView in the ROM, because they have advantages only when that is loaded into memory.
> Can you make an example?
> If there is no Bromite SystemWebView, this will also do nothing?

I don't think I can answer you now in a decent way. 
to answer you fully I should (re)look at that code, I will do it.

**Comment by uazo** [comment]
> I'm not on Note 9, but on a Samsung phone. Bromite is crashing after updating to v99.

have you tried with the corresponding version of chromium? 
in case open an issue with your details.

**Comment by uazo** [comment]
right now this happens to me:

```
03-15 12:01:37.560 12077 12077 W System.err: android.content.res.Resources$NotFoundException: Resource ID #0x0
03-15 12:01:37.560 12077 12077 W System.err:    at android.content.res.ResourcesImpl.getValueForDensity(ResourcesImpl.java:225)
03-15 12:01:37.560 12077 12077 W System.err:    at android.content.res.Resources.getDrawableForDensity(Resources.java:887)
03-15 12:01:37.560 12077 12077 W System.err:    at android.content.res.Resources.getDrawable(Resources.java:827)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.base.ApiCompatibilityUtils.getDrawableForDensity(ApiCompatibilityUtils.java:364)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.base.ApiCompatibilityUtils.getDrawable(ApiCompatibilityUtils.java:333)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.chrome.browser.toolbar.menu_button.MenuButton.updateImageResources(MenuButton.java:136)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.chrome.browser.toolbar.menu_button.MenuButton.showAppMenuUpdateBadge(MenuButton.java:178)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.chrome.browser.toolbar.menu_button.MenuButtonViewBinder.bind(MenuButtonViewBinder.java:43)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.chrome.browser.toolbar.menu_button.MenuButtonViewBinder.bind(MenuButtonViewBinder.java:15)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.ui.modelutil.PropertyModelChangeProcessor.onPropertyChanged(PropertyModelChangeProcessor.java:89)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.ui.modelutil.PropertyModelChangeProcessor.$r8$lambda$24PR1_E-HzQeZ5B9-6klYXNTWZk(Unknown Source:0)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.ui.modelutil.PropertyModelChangeProcessor$$ExternalSyntheticLambda0.onPropertyChanged(Unknown Source:2)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.ui.modelutil.PropertyObservable.notifyPropertyChanged(PropertyObservable.java:68)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.ui.modelutil.PropertyModel.set(PropertyModel.java:412)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.chrome.browser.toolbar.menu_button.MenuButtonMediator.showAppMenuUpdateBadge(MenuButtonMediator.java:224)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.chrome.browser.toolbar.menu_button.MenuButtonMediator.updateStateChanged(MenuButtonMediator.java:210)
03-15 12:01:37.560 12077 12077 W System.err:    at org.chromium.chrome.browser.toolbar.menu_button.MenuButtonCoordinator$$ExternalSyntheticLambda0.run(Unknown Source:2)
03-15 12:01:37.561 12077 12077 W System.err:    at org.chromium.chrome.browser.omaha.UpdateMenuItemHelper.pingObservers(UpdateMenuItemHelper.java:360)
03-15 12:01:37.561 12077 12077 W System.err:    at org.chromium.chrome.browser.omaha.UpdateMenuItemHelper.lambda$new$0$org-chromium-chrome-browser-omaha-UpdateMenuItemHelper(UpdateMenuItemHelper.java:73)
03-15 12:01:37.561 12077 12077 W System.err:    at org.chromium.chrome.browser.omaha.UpdateMenuItemHelper$$ExternalSyntheticLambda2.onResult(Unknown Source:4)
03-15 12:01:37.561 12077 12077 W System.err:    at org.chromium.chrome.browser.omaha.UpdateStatusProvider.pingObservers(UpdateStatusProvider.java:272)
03-15 12:01:37.561 12077 12077 W System.err:    at org.chromium.chrome.browser.omaha.UpdateStatusProvider.resolveStatus(UpdateStatusProvider.java:311)
03-15 12:01:37.561 12077 12077 W System.err:    at org.chromium.chrome.browser.omaha.UpdateStatusProvider.$r8$lambda$SKgNSVpwURZufQR0CY0HyNff5iM(Unknown Source:0)
03-15 12:01:37.561 12077 12077 W System.err:    at org.chromium.chrome.browser.omaha.UpdateStatusProvider$$ExternalSyntheticLambda0.run(Unknown Source:2)
03-15 12:01:37.561 12077 12077 W System.err:    at android.os.MessageQueue.nativePollOnce(Native Method)
03-15 12:01:37.561 12077 12077 W System.err:    at android.os.MessageQueue.next(MessageQueue.java:326)
03-15 12:01:37.561 12077 12077 W System.err:    at android.os.Looper.loop(Looper.java:160)
03-15 12:01:37.561 12077 12077 W System.err:    at android.app.ActivityThread.main(ActivityThread.java:6669)
03-15 12:01:37.561 12077 12077 W System.err:    at java.lang.reflect.Method.invoke(Native Method)
03-15 12:01:37.561 12077 12077 W System.err:    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
03-15 12:01:37.561 12077 12077 W System.err:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```
it seems related to omaha changes

**Comment by uazo** [comment]
@siddhpant if you want to try if it is the same for you, as workaround, longpress in the bromite icon, open a new session incognito and then from the settings disable the check for updates.

**Comment by siddhpant** [comment]
> @siddhpant if you want to try if it is the same for you, as workaround, longpress in the bromite icon, open a new session incognito and then from the settings disable the check for updates.
> ~ @uazo 

Yes, changing the setting does stop the issue.

**Comment by xxxcrow** [comment]
Hey, having crashes on startup xiaomi mi 9 👋
Very strange thing that the problem persists even after clearing the data 😦
I have logs, but can't upload them 😐😑

**Comment by xxxcrow** [comment]
http://paste.c-net.org/SlippersMeans
Sorry I didn't find a better place to upload it 😢 tell me where I can do this

**Comment by drk002** [comment]
Poco Phone X3 - crash app on startup
Poco Phone X3 Pro - Bromite works fine.

Please write how to get the logs then I will paste here.

**Comment by MGCommander** [comment]
OnePlus 8T - crash app on Startup too

**Comment by AlirezaBaloochi** [comment]
Poco F1 with PE ROM - Android 11, same issue.
Clearing Cache and Data doesn't solve anything.
The only fix is reinstalling which works for a few days until it breaks again.

**Comment by csagan5** [comment]
@uazo would it work if it were a command-line argument instead?

**Comment by uazo** [comment]
> would it work if it were a command-line argument instead?

the commandline intended as an object of the chromium framework I think is instantiated later but it would not be a problem. 
But what would a switch be for? and how would the user set it up?

**Comment by csagan5** [comment]
It is explained here: https://www.chromium.org/developers/how-tos/run-chromium-with-flags/

It would be more neutral than looking up device model and allow usage for different models. It seems that security-wise the change of linker should be fine as it only affects performance, but I did not quite understand how Bromite uses RELRO, maybe only when the SystemWebView is also installed?

**Comment by csagan5** [comment]
@AlirezaBaloochi you do not have a Note9 so this issue does not apply to your device.

**Comment by uazo** [comment]
> As far as I know monochrome or trichrome have 0 advantages if you do not also modify the SystemWebView in the ROM, because they have advantages only when that is loaded into memory.

seems to be correct. the advantage between Modern, Legacy and System linker seems to be in the amount of memory used for the shared relro _with the webview_, and is active only for monochrome and trichrome where it makes sense.

some excerpt:
```
    /**
     * Returns either a LegacyLinker or a ModernLinker.
     *
     * ModernLinker requires OS features from Android M and later: a system linker that handles
     * packed relocations and load from APK, and |android_dlopen_ext()| for shared RELRO support. It
     * cannot run on Android releases earlier than M.
     *
     * LegacyLinker runs on all Android releases but it is slower and more complex than
     * ModernLinker. The LegacyLinker is used on M as it avoids writing the relocation to disk.
     *
     * On N, O and P Monochrome is selected by Play Store. With Monochrome this code is not used,
     * instead Chrome asks the WebView to provide the library (and the shared RELRO). If the WebView
     * fails to provide the library, the system linker is used as a fallback.
     *
     * LegacyLinker can run on all Android releases, but is unused on P+ as it may cause issues.
     * LegacyLinker is preferred on M- because it does not write the shared RELRO to disk at
     * almost every cold startup.
     *
     * Finally, ModernLinker is used on Android Q+ with Trichrome.
     *
     * More: docs/android_native_libraries.md
     *
     * @return the Linker implementation instance.
     */
```

```
LegacyLinker is buggy on Android 10 so fallback to system linker on this platform
```

```
        // A non-monochrome APK (such as ChromePublic.apk) can be installed on N+ in these
        // circumstances:
        // * installing APK manually
        // * after OTA from M to N
        // * side-installing Chrome (possibly from another release channel)
        // * Play Store bugs leading to incorrect APK flavor being installed
        // * installing other Chromium-based browsers
        //
        // For Chrome builds regularly shipped to users on N+, the system linker (or the Android
        // Framework) provides the necessary functionality to load without crazylinker. The
        // LegacyLinker is risky to auto-enable on newer Android releases, as it may interfere with
        // regular library loading. See http://crbug.com/980304 as example.
        //
        // This is only called if LibraryLoader.useChromiumLinker() returns true, meaning this is
        // either Chrome{,Modern} or Trichrome.
```

> It seems that security-wise the change of linker should be fine as it only affects performance, 

memory occupation

> but I did not quite understand how Bromite uses RELRO, maybe only when the SystemWebView is also installed?

yes, is my own conclusion

**Comment by csagan5** [comment]
Ok; then I would accept a patch for this via command-line arguments.

**Comment by uazo** [comment]
> I would accept a patch for this via command-line arguments.

I'm sorry but it's not possible. 
the command line in android is not manageable by the user, and in the production build it is necessary to set a flag to be able to manage the artifice of "`/data/local/tmp/chrome-command-line`", but, as it goes into crash on startup, the flag cannot be set

**Comment by csagan5** [comment]
> > I would accept a patch for this via command-line arguments.
> 
> I'm sorry but it's not possible.

I guess [here](https://github.com/bromite/bromite/issues/1871#issuecomment-1081459771) you were referring to a different command-line then.

Merging the patch for this issue is problematic because we do not have a device to test this so it could stay in Bromite forever; what happens if it creates other problems on that device? In general it looks like we do not use RELRO at all unless the SystemWebView is also installed so at least it should not introduce security issues.

@uazo please submit a PR also for this patch, although this decision needs to be re-considered if there are more devices with this problem

**Comment by csagan5** [comment]
> > what happens if it creates other problems on that device?
> 
> more than crash on startup?

Yes, for example with calls to native Android (for video playback etc); all bug reports for Note9 will have to consider this fundamental linker change.

**Comment by uazo** [comment]
> all bug reports for Note9 will have to consider this fundamental linker change.

I would say no, being the fallback in case crazy doesn't work, and it could be the same for other devices but without crash

**Comment by csagan5** [comment]
I have added the patch for next release.

**Comment by csagan5** [comment]
Fixed in `100.0.4896.83`.
